### PR TITLE
API-2532: Addition of fields definition endpoint and subscription fie…

### PR DESCRIPTION
**Pull request by googley42**
…lds bulk GET endpoint

### CODE REVIEW

File: app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala [hunk 1]
```diff
@@ -72,19 +76,28 @@ class FieldsDefinitionMongoRepository @Inject()(mongoDbProvider: MongoDbProvider
...
-  override def save(fieldsDefinition: FieldsDefinition): Future[Unit] = {
-    val selector = Json.obj("id" -> fieldsDefinition.id)
-    Logger.debug(s"[save] selector: $selector")
-    collection.find(selector).one[BSONDocument].flatMap {
-      case Some(document) => collection.update(selector = BSONDocument("_id" -> document.get("_id")), update = fieldsDefinition)
-      case _ => collection.insert(fieldsDefinition)
-    }.map {
-      writeResult => handleError(writeResult, s"Could not save fields definition fields: $fieldsDefinition")
+  /**
+    * Saves or inserts entity depending on if it already exists.
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala]: I would remove the Scaladoc.
The code is simple and the reader can understand what the method does.

File: app/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionService.scala [hunk 2]
```diff
@@ -29,22 +29,9 @@ import scala.concurrent.Future
...
 
-    repository.fetchById(identifier.encode()) flatMap {
-      o =>
-        o.fold(
-          create() map { _ => true }
-        )(
-          existing => update(existing.id) map { _ => false }
-        )
-    }
+    repository.upsert(FieldsDefinition(identifier.encode(), fields)) map (isInserted => isInserted)
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionService.scala]: What is the purpose of adding `map ( x => x )` here?
Does it do anything useful?
For me it seems useless.
You are mapping something to itself, without any transformation.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 3]
```diff
@@ -74,18 +84,23 @@ class SubscriptionFieldsMongoRepository @Inject()(mongoDbProvider: MongoDbProvid
...
-  override def save(subscription: SubscriptionFields): Future[Unit] = {
-    val selector = selectorById(subscription.id)
-    Logger.debug(s"[save] selector: $selector")
-    collection.find(selector).one[BSONDocument].flatMap {
-      case Some(document) => collection.update(selector = BSONDocument("_id" -> document.get("_id")), update = subscription)
-      case _ => collection.insert(subscription)
-    }.map {
-      writeResult => handleError(writeResult, s"Could not save subscription fields: $subscription")
+  /**
+    * Saves or inserts entity depending on if it already exists.
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Remove scaladoc, we do not use it anywhere.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 4]
```diff
@@ -74,18 +84,23 @@ class SubscriptionFieldsMongoRepository @Inject()(mongoDbProvider: MongoDbProvid
...
+    * Returns Future of isInserted Boolean flag if everything went OK - otherwise a failed Future with an error message.
+    * @param subscription entity to upsert
+    */
+  def upsert(subscription: SubscriptionFields): Future[Boolean] = {
+    collection.update(selector = BSONDocument("_id" -> subscription.id), update = subscription, upsert = true).map {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Why did you change the code here?

Also, are you sure this code is correct?
`_id` is the mongo primary key.
`id` is the field we defined.
They are two different things.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala [hunk 5]
```diff
@@ -72,19 +76,28 @@ class FieldsDefinitionMongoRepository @Inject()(mongoDbProvider: MongoDbProvider
...
+    * Returns Future of isInserted Boolean flag if everything went OK - otherwise a failed Future with an error message.
+    * @param fieldsDefinition entity to upsert
+    */
+  def upsert(fieldsDefinition: FieldsDefinition): Future[Boolean] = {
+    collection.update(selector = BSONDocument("_id" -> fieldsDefinition.id), update = fieldsDefinition, upsert = true).map {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala]: `_id` is the mongo primary key.
`id` is the field we defined.
They are two different things.

I suspect this code might have something wrong.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 6]
```diff
@@ -36,9 +37,14 @@ import scala.concurrent.Future
 @ImplementedBy(classOf[SubscriptionFieldsMongoRepository])
 trait SubscriptionFieldsRepository {
 
-  //TODO consider Future[Boolean]
-  def save(subscription: SubscriptionFields): Future[Unit]
+  /**
+    * Saves or inserts entity depending on if it already exists.
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: Do not worry about scaladoc. 
Let's remove it from anywhere.
If the code does something not easy to read or you need to add some notes, you can add a comment above the instruction.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala [hunk 7]
```diff
@@ -72,19 +76,28 @@ class FieldsDefinitionMongoRepository @Inject()(mongoDbProvider: MongoDbProvider
...
+      updateWriteResult => handleError(updateWriteResult, "", updateWriteResult.upserted.nonEmpty)
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala]: Why empty string?

**Comment by gokyo** [comment]
`FieldsDefinitionRepository` and `SubscriptionFieldsRepository` have several code in common.
Could you extract it, write it only once, and re-use it?

**Comment by gokyo** [comment]
I am happy to use the method name `upsert` in the service and controller layers.
But in the repository layer, let's use the method `save` instead of `upsert`.

**Comment by gokyo** [comment]
Could you rename `InMemoryRepository` to `SubscriptionFieldsInMemoryRepository`?
I already commented this in https://github.com/hmrc/api-subscription-fields/pull/8

**Comment by googley42** [comment]
re comment `FieldsDefinitionRepository and SubscriptionFieldsRepository have several code in common.
Could you extract it, write it only once, and re-use it?` we are planning another PR with generic repo code. We did not address here to keep the PR size down. Also in the next PR we thinking of removing in memory all In Memory Repo's are they are not used. We mock out repo's in service unit tests, and use real Mongo in repo integration tests

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 8]
```diff
@@ -74,18 +84,23 @@ class SubscriptionFieldsMongoRepository @Inject()(mongoDbProvider: MongoDbProvid
...
+    * Returns Future of isInserted Boolean flag if everything went OK - otherwise a failed Future with an error message.
+    * @param subscription entity to upsert
+    */
+  def upsert(subscription: SubscriptionFields): Future[Boolean] = {
+    collection.update(selector = BSONDocument("_id" -> subscription.id), update = subscription, upsert = true).map {
+      updateWriteResult => handleUpsertError(updateWriteResult, s"Could not save subscription fields: $subscription", updateWriteResult.upserted.nonEmpty)
     }
   }
 
+  def fetchByApplicationId(applicationId: String): Future[List[SubscriptionFields]] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: you forgot `override` in front

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala [hunk 9]
```diff
@@ -74,18 +84,23 @@ class SubscriptionFieldsMongoRepository @Inject()(mongoDbProvider: MongoDbProvid
...
-      case Some(document) => collection.update(selector = BSONDocument("_id" -> document.get("_id")), update = subscription)
-      case _ => collection.insert(subscription)
-    }.map {
-      writeResult => handleError(writeResult, s"Could not save subscription fields: $subscription")
+  /**
+    * Saves or inserts entity depending on if it already exists.
+    * Returns Future of isInserted Boolean flag if everything went OK - otherwise a failed Future with an error message.
+    * @param subscription entity to upsert
+    */
+  def upsert(subscription: SubscriptionFields): Future[Boolean] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: you forgot `override` in front

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala [hunk 10]
```diff
@@ -21,7 +21,7 @@ import java.util.UUID
 import uk.gov.hmrc.apisubscriptionfields.model.{Fields, SubscriptionIdentifier}
 
 object SubscriptionFields {
-  def apply(id: SubscriptionIdentifier, fieldsId: UUID, customFields: Fields) : SubscriptionFields = SubscriptionFields(id.encode(), fieldsId, customFields)
+  def apply(id: SubscriptionIdentifier, fieldsId: UUID, customFields: Fields) : SubscriptionFields = new SubscriptionFields(id.encode(), id.applicationId.value, fieldsId, customFields)
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: could you rename `customFields` to `fields`?

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala [hunk 11]
```diff
@@ -21,7 +21,7 @@ import java.util.UUID
...
 }
 
-case class SubscriptionFields(id: String, fieldsId: UUID, customFields: Fields)
+case class SubscriptionFields(id: String, applicationId: String, fieldsId: UUID, customFields: Fields)
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: The idea of adding `applicationId` as additional field and index in this mongo repository is a good idea.
This in fact allow you to find efficiently records by `applicationId`.

However, if you follow this approach, you will have repetitions in `id` and `applicationId`, as `id` contains the `applicationId`.

With this change, I think it is best to remove the `id` field, and introduce a new field called `apiIdentifier`.
The value of the field `apiIdentifier` could be `<apiContext>__<apiVersion>`.

In this mongo repository we would need:
- an index for the single field `applicationId`
- an index for the two fields `applicationId` and `apiIdentifier`

What do you think about this optimisation?

Obviously this change will need a some refactoring everywhere in the code, but I believe this is the way to go.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala [hunk 12]
```diff
@@ -35,8 +35,12 @@ import scala.concurrent.Future
...
-  //TODO consider Future[Boolean]
-  def save(fieldsDefinition: FieldsDefinition): Future[Unit]
+  /**
+    * Saves or inserts entity depending on if it already exists.
+    * Returns Future of isInserted Boolean flag if everything went OK - otherwise a failed Future with an error message.
+    * @param fieldsDefinition entity to upsert
+    */
+  def upsert(fieldsDefinition: FieldsDefinition): Future[Boolean]
 
   def fetchById(id: String): Future[Option[FieldsDefinition]]
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala]: I would rename it to:
```
fetch(apiIdentifier: String)
```

**Comment by gokyo** [comment]
Could you change `uk.gov.hmrc.apisubscriptionfields.repository.FieldsDefinition` to the following?

```
object FieldsDefinition {
  def apply(apiIdentifier: ApiIdentifier, fields: Seq[FieldDefinition]) : FieldsDefinition = FieldsDefinition(apiIdentifier, fields)
}

case class FieldsDefinition(iapiIdentifier: ApiIdentifier, fields: Seq[FieldDefinition])
```

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala (hunk 10 above)
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: Found an empty space.
Rewrite to 
```
def apply(id: SubscriptionIdentifier, fieldsId: UUID, customFields: Fields): SubscriptionFields = .......
```

File: conf/app.routes [hunk 13]
```diff
@@ -1,17 +1,18 @@
 # microservice specific routes
 
-
 # subscription field definitions endpoints routes
+PUT           /definition/context/:apiContext/version/:apiVersion                      @uk.gov.hmrc.apisubscriptionfields.controller.FieldsDefinitionController.upsertFieldsDefinition(apiContext: String, apiVersion: String)
 
-# TODO
+GET           /definition/context/:apiContext/version/:apiVersion                      @uk.gov.hmrc.apisubscriptionfields.controller.FieldsDefinitionController.getFieldsDefinition(apiContext: String, apiVersion: String)
```
Comment (gokyo) [inline conf/app.routes]: Could you please add a new route?
```
GET     /definition/
```

This would need to return a JSON with all field definitions

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala (hunk 11 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: The original `id` field is an encoded one of format `{application-id}___{api-context}___{api-version}` the encode process may add extra separator chars in first part of the above key. This means that the index on the above key can not be used for plain `applicationId` eg regex match using `^appId` (as indexed field may contain extra separator chars). So that's why we introduced the secondary `applicationId` index. My understanding is that we *both* indexes as we still need to do single record operations using `{application-id}___{api-context}___{api-version}` as well as bulk operations using `applicationId`

File: conf/app.routes [hunk 14]
```diff
@@ -1,17 +1,18 @@
...
+GET           /field/application/:appId/context/:apiContext/version/:apiVersion        @uk.gov.hmrc.apisubscriptionfields.controller.ApiSubscriptionFieldsController.getSubscriptionFields(appId: String, apiContext: String, apiVersion: String)
 
-GET         /field/application/:appId/context/:apiContext/version/:apiVersion        @uk.gov.hmrc.apisubscriptionfields.controller.ApiSubscriptionFieldsController.getSubscriptionFields(appId: String, apiContext: String, apiVersion: String)
+GET           /field/:fieldsId                                                         @uk.gov.hmrc.apisubscriptionfields.controller.ApiSubscriptionFieldsController.getSubscriptionFieldsByFieldsId(fieldsId: java.util.UUID)
 
-GET         /field/:fieldsId                                                         @uk.gov.hmrc.apisubscriptionfields.controller.ApiSubscriptionFieldsController.getSubscriptionFieldsByFieldsId(fieldsId: java.util.UUID)
+GET           /field/application/:appId                                                @uk.gov.hmrc.apisubscriptionfields.controller.ApiSubscriptionFieldsController.getBulkSubscriptionFieldsByApplicationId(appId: String)
```
Comment (gokyo) [inline conf/app.routes]: Could you please add a new route?
```
GET     /field/
```

This would need to return a JSON with all subscription fields.

I know we do not use it anywhere in the code, but this would be very useful in case of investigations in production for example, where we do not have access to mongoDb.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala (hunk 11 above)
Comment (gokyo) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: At the moment the separator is `##`, not `__`.
Thus, the `id` field contains `{applicationId}___{apiContext}___{apiVersion}`.

I know we need the index on the field `applicationId`, as you need it when you fetch by `applicationId`.

But now the field `id` could be removed.
If you have the fields `applicationId` and `apiIdentifier` (where `apiIdentifier` is `{apiContext}___{apiVersion}`), you can introduce two mongo indexes:
1) one index on the single field `applicationId`
2) one index on the fields `applicationId` and `apiIdentifier`

When you fetch subscription fields using both the application id and the API (context and version), you need to use a selector on both `applicationId` and `apiIdentifier`.
This fetch operation will use the second index mentioned above.

We do not need to split `apiIdentifier` in two fields.
You can have both in the same field, using a separator (for example `__` or `##`).

Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: OK got it. Sounds like a good idea but a fair deal of work. Maybe a candidate for another (next) PR?

Comment (gokyo) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: If there is something new to add, I am happy for the work to be done in a new PR.
But since this comment is about code you committed in this PR, I think it is best to adjust it here in this PR, before to merge to master.

File: app/uk/gov/hmrc/apisubscriptionfields/model/Responses.scala [hunk 15]
```diff
@@ -19,8 +19,9 @@ package uk.gov.hmrc.apisubscriptionfields.model
 import play.api.libs.json.Json.JsValueWrapper
 import play.api.libs.json.{JsObject, Json}
 
-case class SubscriptionFieldsIdResponse(fieldsId: SubscriptionFieldsId)
-case class SubscriptionFieldsResponse(fieldsId: SubscriptionFieldsId, fields: Fields)
+case class SubscriptionFieldsResponse(id: String, fieldsId: SubscriptionFieldsId, fields: Fields)
+
+case class BulkSubscriptionFieldsResponse(fields: Seq[SubscriptionFieldsResponse])
 
 case class FieldsDefinitionResponse(fields: Seq[FieldDefinition])
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/model/Responses.scala]: for avoiding confusion, I suggest to rename `fields` to `fieldDefinitions`

File: app/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionService.scala [hunk 16]
```diff
@@ -29,22 +29,9 @@ import scala.concurrent.Future
 class FieldsDefinitionService @Inject() (repository: FieldsDefinitionRepository) {
 
   def upsert(identifier: FieldsDefinitionIdentifier, fields: Seq[FieldDefinition]): Future[Boolean] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionService.scala]: for avoiding confusion, I suggest to rename `fields` to `fieldDefinitions`

**Comment by gokyo** [comment]
@paulblakehmrc & @googley42 
I have reviewed again this.

I think we could get rid completely of the encoding/decoding parts everywhere.

For the `subscription fields` mongo repository, I suggest:
these fields: 

- `applicationId`
- `apiContext`
- `apiVersion`
- `fieldsId`
- `fields`

and these indexes:

- index on the single field `fieldsId`
- index on the single field `applicationId`
- index on these 3 fields: `applicationId`, `apiContext`, `apiVersion`


For the `field definitions` mongo repository, I suggest:
these fields: 

- `apiContext`
- `apiVersion`
- `fieldDefinitions`

and this index:

- index on these 2 fields: `apiContext`, `apiVersion`

**Comment by googley42** [comment]
Great! this what I wanted. `{applicationId}<sep_chars>{apiContext}<sep_chars>{apiVersion}` was an internal detail and used by the client. Getting rid of encoding and decoding will *vastly* simplify the code.

**Comment by gokyo** [comment]
I close this PR, as we agreed it won't be merged to `master`.
I am looking forward to see another PR including the changes asked here.

**Comment by gokyo** [comment]
@googley42 
If you see other improvements, tell us, do not wait for us to find them, as it happened now.

File: app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala (hunk 1 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala]: done

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 3 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: done

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 6 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: done

File: app/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionService.scala (hunk 2 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/service/FieldsDefinitionService.scala]: done

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 9 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: done

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala (hunk 10 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: done

Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFields.scala]: done

File: app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala (hunk 4 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/SubscriptionFieldsRepository.scala]: done

File: app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala (hunk 7 above)
Comment (googley42) [reply app/uk/gov/hmrc/apisubscriptionfields/repository/FieldsDefinitionRepository.scala]: done
